client/pluginmanager/csimanager/volume_test.go

```diff
@@ -597,3 +598,113 @@ func TestVolumeManager_InterleavedStaging(t *testing.T) {
 	must.Eq(t, 1, csiFake.NodeUnpublishVolumeCallCount, must.Sprint("expected 1 unpublish call"))
 	must.Eq(t, 0, csiFake.NodeUnstageVolumeCallCount, must.Sprint("expected no unstage call"))
 }
+
+func TestVolumeManager_Serialization(t *testing.T) {
+	ci.Parallel(t)
+
+	tmpPath := t.TempDir()
+	csiFake := &csifake.Client{}
+
+	logger := testlog.HCLogger(t)
+
+	ctx := hclog.WithContext(t.Context(), logger)
+
+	manager := newVolumeManager(logger,
+		func(e *structs.NodeEvent) {}, csiFake,
+		tmpPath, tmpPath, true, "i-example")
+
+	// ensure that serialized ops for the same volume cannot interleave.
+
+	ctx, cancel := context.WithTimeout(t.Context(), time.Second)
+	t.Cleanup(cancel)
+	errs := make(chan error, 3)
+	results := make(chan string, 3)
+
+	// we can't guarantee the goroutines will contend, so we'll force the first
+	// op to wait until the other goroutines have at least started, and then
+	// sleep for a bit to give them a chance to catch up
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			wg.Wait()
+			time.Sleep(100 * time.Millisecond)
+			results <- "one"
+			return errors.New("one")
+		})
+	}()
+
+	go func() {
+		wg.Done() // unblock the first op
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			results <- "two"
+			return errors.New("two")
+		})
+	}()
+
+	go func() {
+		wg.Done() // unblock the first op
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			results <- "three"
+			return errors.New("three")
+		})
+	}()
+
+	found := 0
+	for {
+		if found >= 3 {
+			break
+		}
+		select {
+		case err := <-errs:
+			found++
+			select {
+			case result := <-results:
+				if err.Error() != result {
+					t.Fatal("got result out-of-order from error")
+				}
+			case <-ctx.Done():
+				t.Fatal("did not get result for error before timeout")
+			}
+		case <-ctx.Done():
+			t.Fatal("timed out waiting for error")
+		}
+
+	}
+
+	// ensure that serialized ops for the different volumes can interleave
+	wg.Add(1)
+
+	go func() {
+		errs <- manager.serializedOp(ctx, "ns", "vol0", func() error {
+			wg.Wait()
+			time.Sleep(100 * time.Millisecond)
+			return errors.New("four")
+		})
+	}()
+
+	go func() {
+		errs <- manager.serializedOp(ctx, "ns", "vol1", func() error {
+			wg.Done() // unblock the first op
+			return errors.New("five")
+		})
+	}()
+
+	ctx2, cancel2 := context.WithTimeout(t.Context(), time.Second)
+	t.Cleanup(cancel2)
+
+	found = 0
+	for {
+		if found >= 2 {
+			break
+		}
+		select {
+		case <-errs:
+			found++
+		case <-ctx2.Done():
+			t.Fatal("timed out waiting for error")
+		}
+	}
```

Comment:
I don't see how this bottom case that proves that they could interleave, just that they both ran (within 1 second).

Reply:
It's a little subtle, but the `wait.Done` in the second goroutine is located _inside_ the serialization function (as opposed to outside, as it is in the first test case). That means that we can't run the first one at all unless the second one is also running.

Reply:
it only means that the first can't return until the second has called `wg.Done`, but that doesn't mean that it's not also after the second one's `return`. 

one way to check this empirically would be to drop `serializedOp`.


```go
	go func() {
		errs <- func() error {
			wg.Wait()
			time.Sleep(100 * time.Millisecond)
			return errors.New("four")
		}()
	}()

	go func() {
		errs <- func() error {
			wg.Done() // unblock the first op
			return errors.New("five")
		}()
	}()
```

I could be mistaken, but I suspect that this would still pass, because they would both run within the `time.Second`

Reply:
as written, the only indication we have of whether they are running or not is whether they have returned. we do not know whether they ran at the same time.

I'm not sure of a better way to get at this, but if they both slept for 100ms, and the total duration was less than 200ms, then we could be confident that they ran simultaneously.

Reply:
Ah, I see what you're saying. If the second goroutine runs to completion before we even start the first goroutine, the first goroutine will run and then immediately unblock. I think we can fix that with a second wait group going the other direction.

Reply:
hey yeah that does it!

and we can save some milliseconds now, don't need that `time.Sleep` in the first func.